### PR TITLE
client/beefy: fix on-demand justifications sync for old blocks

### PR DESCRIPTION
**Pull request by acatangiu**
When receiving BEEFY justifications for old blocks the state might be pruned for them, in which case justification verification fails because BEEFY validator set cannot be retrieved from runtime state.

Fix this by having the voter give the validator set to the `OnDemandJustificationsEngine` as request information. On receiving a BEEFY justification for requested block, the provided validator set will be used to validate the justification.

### CODE REVIEW

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2078057

**Comment by acatangiu** [comment]
bot rebase

**Review by svyatonik — APPROVED** [review]
LGTM

File: client/beefy/src/worker.rs [hunk 1]
```diff
@@ -167,12 +166,12 @@ impl<B: Block> VoterOracle<B> {
 	}
 
 	/// Return current pending mandatory block, if any.
```
Comment (bkchr) [inline client/beefy/src/worker.rs:168]: Stale docs.

File: client/beefy/src/worker.rs [hunk 2]
```diff
@@ -551,10 +548,16 @@ where
...
-				if self_vote || self.voting_oracle().mandatory_pending() == Some(round.1) {
-					// Persist state after handling self vote to avoid double voting in case
-					// of voter restarts.
-					// Also persist state after handling mandatory block vote.
+				let mandatory_round =
+					if let Some((number, _)) = self.voting_oracle().mandatory_pending() {
+						number == round.1
+					} else {
+						false
+					};
```
Comment (bkchr) [inline client/beefy/src/worker.rs:556]:
```suggestion
					self.voting_oracle().mandatory_pending().map(|p| p.0 == round.1).unwrap_or(false);
```

File: client/beefy/src/communication/request_response/outgoing_requests_engine.rs [hunk 3]
```diff
@@ -100,10 +94,13 @@ where
 		None
 	}
 
-	fn request_from_peer(&mut self, peer: PeerId, block: NumberFor<B>) {
-		debug!(target: "beefy::sync", "🥩 requesting justif #{:?} from peer {:?}", block, peer);
+	fn request_from_peer(&mut self, peer: PeerId, req_info: RequestInfo<B>) {
+		debug!(
+			target: "beefy::sync", "🥩 requesting justif #{:?} from peer {:?}",
+			req_info.block, peer
+		);
```
Comment (bkchr) [inline client/beefy/src/communication/request_response/outgoing_requests_engine.rs:101]:
```suggestion
		debug!(
			target: "beefy::sync", 
			"🥩 requesting justif #{:?} from peer {:?}",
			req_info.block, 
			peer,
		);
```

**Review by bkchr — APPROVED** [review]

**Pull request merged**